src/EditorFeatures/CSharp/MisplacedUsingDirectives/MisplacedUsingDirectivesCodeFixProvider.cs

```diff
@@ -0,0 +1,394 @@
+
+            // Do not offer a code fix when there are multiple namespaces in the source file.
+            if (CountNamespaces(compilationUnit.Members) > 1)
+            {
+                return;
+            }
+
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                context.RegisterCodeFix(
+                    new MoveMisplacedUsingsCodeAction(cancellationToken => GetTransformedDocumentAsync(context.Document, compilationUnit, options, cancellationToken)),
+                    diagnostic);
+            }
+        }
+
+        private static async Task<Document> GetTransformedDocumentAsync(Document document, CompilationUnitSyntax compilationUnit, OptionSet options, CancellationToken cancellationToken)
+        {
+            var usingDirectivesPlacement = DeterminePlacement(compilationUnit, options);
+
+            // There should only be a diagnostic when there are usings directives that need moving.
+            Debug.Assert(!usingDirectivesPlacement.preferPreservation);
+
+            var compilationUnitWithExpandedUsings = await ExpandUsingDirectivesAsync(document, compilationUnit, cancellationToken).ConfigureAwait(false);
+
+            var (compilationUnitWithoutHeader, fileHeader) = RemoveFileHeader(compilationUnitWithExpandedUsings);
+
+            var newCompilationUnit = usingDirectivesPlacement.placement switch
+            {
+                AddImportPlacement.InsideNamespace => MoveUsingsInsideNamespace(compilationUnitWithoutHeader),
+                AddImportPlacement.OutsideNamespace => MoveUsingsOutsideNamespace(compilationUnitWithoutHeader),
+                _ => throw ExceptionUtilities.Unreachable
+            };
+
+            var newCompilationUnitWithHeader = AddFileHeader(newCompilationUnit, fileHeader);
+
+            var newDocument = document.WithSyntaxRoot(newCompilationUnitWithHeader);
+
+            return await Simplifier.ReduceAsync(newDocument, Simplifier.Annotation, options, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<CompilationUnitSyntax> ExpandUsingDirectivesAsync(Document document, CompilationUnitSyntax compilationUnit, CancellationToken cancellationToken)
+        {
+            // Expand the usings declarations in the compilation unit.
+            var expandCompilationUnitUsingsTasks = compilationUnit.Usings.Select(
+                usingDirective => ExpandUsingDirectiveAsync(document, usingDirective, cancellationToken));
+            var newCompilationUnitUsings = await Task.WhenAll(expandCompilationUnitUsingsTasks).ConfigureAwait(false);
+
+            var compilationUnitWithExpandedUsings = compilationUnit.WithUsings(new SyntaxList<UsingDirectiveSyntax>(newCompilationUnitUsings));
+
+            // To expand the using declarations in the namespace we need to get an updated document, since the 
+            // expander requires that the SyntaxTree for the semantic model match that of the using directives.
```

Comment:
my recommendation in cases like this is to use a new method and only pass in the newDocument.  That way you don't have to worry about accidentally using the old versus new state.